File: services/storefront/app/Http/Controllers/AuthController.php
<?php

namespace App\Http\Controllers;

use App\Services\ApiGateway;
use Illuminate\Http\Request;

class AuthController extends Controller
{
    private ApiGateway $api;

    public function __construct(ApiGateway $api)
    {
        $this->api = $api;
    }

    public function showLogin()
    {
        if (session('auth_token')) {
            return redirect('/');
        }
        // Store the previous page URL so we can redirect back after login
        $intended = session('redirect') ?? url()->previous();
        if (str_contains($intended, '/login') || str_contains($intended, '/register')) {
            $intended = '/';
        }
        session(['url.intended' => $intended]);
        return view('pages.login', ['title' => 'Login - ZylkerKart', 'intended' => $intended]);
    }

    public function showRegister()
    {
        if (session('auth_token')) {
            return redirect('/');
        }
        return view('pages.register', ['title' => 'Register - ZylkerKart']);
    }

    public function login(Request $request)
    {
        $request->validate([
            'email'    => 'required|email',
            'password' => 'required|min:6',
        ]);

        $result = $this->api->post('auth', '/auth/login', [
            'email'    => $request->email,
            'password' => $request->password,
        ]);

        if ($result['status'] === 200) {
            session([
                'auth_token'    => $result['data']['accessToken'],
                'refresh_token' => $result['data']['refreshToken'],
                'user'          => $result['data']['user'],
            ]);

            $redirect = $request->get('redirect', session('url.intended', '/'));
            session()->forget('url.intended');
            return redirect($redirect)->with('success', 'Welcome back!');
        }

        return back()->withErrors([
            'email' => $result['data']['error'] ?? 'Login failed',
        ]);
    }

    public function register(Request $request)
    {
        $request->validate([
            'full_name' => 'required|max:255',
            'email'     => 'required|email',
            'password'  => 'required|min:8|confirmed',
            'phone'     => 'nullable|max:20',
        ]);

        $result = $this->api->post('auth', '/auth/register', [
            'email'    => $request->email,
            'password' => $request->password,
            'fullName' => $request->full_name,
            'phone'    => $request->phone,
        ]);

        if ($result['status'] === 200) {
            session([
                'auth_token'    => $result['data']['accessToken'],
                'refresh_token' => $result['data']['refreshToken'],
                'user'          => $result['data']['user'],
            ]);

            return redirect('/')->with('success', 'Account created successfully!');
        }

        return back()->withErrors([
            'email' => $result['data']['error'] ?? 'Registration failed',
        ]);
    }

    public function logout()
    {
        $refreshToken = session('refresh_token');
        if ($refreshToken) {
            $this->api->post('auth', '/auth/logout', [
                'refreshToken' => $refreshToken,
            ]);
        }

        session()->forget(['auth_token', 'refresh_token', 'user']);
        return redirect('/')->with('success', 'Logged out successfully');
    }
}

File: services/storefront/src/main/java/com/zylkerkart/storefront/controller/AuthController.java
package com.zylkerkart.storefront.controller;

import com.zylkerkart.storefront.service.ApiGateway;
import jakarta.servlet.http.HttpSession;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

@Controller
public class AuthController {

    private final ApiGateway api;

    public AuthController(ApiGateway api) {
        this.api = api;
    }

    @GetMapping("/login")
    public String showLogin(Model model, HttpSession session,
                            @RequestParam(value = "redirect", required = false) String redirect,
                            @RequestHeader(value = "Referer", required = false) String referer) {
        if (session.getAttribute("auth_token") != null) {
            return "redirect:/";
        }
        // Store the page to return to after login
        if (redirect != null && !redirect.isEmpty()) {
            session.setAttribute("redirect", redirect);
        } else if (referer != null && !referer.isEmpty() && !referer.contains("/login") && !referer.contains("/register")) {
            // Extract path from referer URL (strip host)
            try {
                java.net.URI uri = new java.net.URI(referer);
                String path = uri.getPath();
                if (path != null && !path.isEmpty()) {
                    session.setAttribute("redirect", path);
                }
            } catch (Exception ignored) {
                // Skip invalid referer
            }
        }
        model.addAttribute("title", "Login - ZylkerKart");
        model.addAttribute("user", session.getAttribute("user"));
        model.addAttribute("sessionId", session.getId());
        model.addAttribute("authToken", session.getAttribute("auth_token"));
        return "pages/login";
    }

    @GetMapping("/register")
    public String showRegister(Model model, HttpSession session) {
        if (session.getAttribute("auth_token") != null) {
            return "redirect:/";
        }
        model.addAttribute("title", "Register - ZylkerKart");
        model.addAttribute("user", session.getAttribute("user"));
        model.addAttribute("sessionId", session.getId());
        model.addAttribute("authToken", session.getAttribute("auth_token"));
        return "pages/register";
    }

    @PostMapping("/login")
    public String login(
            @RequestParam String email,
            @RequestParam String password,
            HttpSession session,
            Model model) {

        Map<String, Object> body = new HashMap<>();
        body.put("email", email);
        body.put("password", password);

        Map<String, Object> result = api.post("auth", "/auth/login", body);
        int status = (int) result.get("status");

        if (status >= 200 && status < 300) {
            @SuppressWarnings("unchecked")
            Map<String, Object> data = (Map<String, Object>) result.get("data");
            if (data != null) {
                session.setAttribute("auth_token", data.get("accessToken"));
                session.setAttribute("user", data.get("user"));
            }
            String redirect = (String) session.getAttribute("redirect");
            session.removeAttribute("redirect");
            return "redirect:" + (redirect != null ? redirect : "/");
        }

        @SuppressWarnings("unchecked")
        Map<String, Object> errorData = (Map<String, Object>) result.get("data");
        String error = errorData != null
                ? (String) errorData.getOrDefault("error", "Invalid email or password")
                : "Invalid email or password";

        model.addAttribute("error", error);
        model.addAttribute("email", email);
        model.addAttribute("title", "Login - ZylkerKart");
        model.addAttribute("user", session.getAttribute("user"));
        model.addAttribute("sessionId", session.getId());
        model.addAttribute("authToken", session.getAttribute("auth_token"));
        return "pages/login";
    }

    @PostMapping("/register")
    public String register(
            @RequestParam String name,
            @RequestParam String email,
            @RequestParam String password,
            @RequestParam(name = "password_confirmation") String passwordConfirmation,
            HttpSession session,
            Model model) {

        if (!password.equals(passwordConfirmation)) {
            model.addAttribute("error", "Passwords do not match");
            model.addAttribute("name", name);
            model.addAttribute("email", email);
            model.addAttribute("title", "Register - ZylkerKart");
            model.addAttribute("user", session.getAttribute("user"));
            model.addAttribute("sessionId", session.getId());
            model.addAttribute("authToken", session.getAttribute("auth_token"));
            return "pages/register";
        }

        Map<String, Object> body = new HashMap<>();
        body.put("name", name);
        body.put("email", email);
        body.put("password", password);

        Map<String, Object> result = api.post("auth", "/auth/register", body);
        int status = (int) result.get("status");

        if (status >= 200 && status < 300) {
            @SuppressWarnings("unchecked")
            Map<String, Object> data = (Map<String, Object>) result.get("data");
            if (data != null) {
                session.setAttribute("auth_token", data.get("accessToken"));
                session.setAttribute("user", data.get("user"));
            }
            return "redirect:/";
        }

        @SuppressWarnings("unchecked")
        Map<String, Object> errorData = (Map<String, Object>) result.get("data");
        String error = errorData != null
                ? (String) errorData.getOrDefault("error", "Registration failed")
                : "Registration failed";

        model.addAttribute("error", error);
        model.addAttribute("name", name);
        model.addAttribute("email", email);
        model.addAttribute("title", "Register - ZylkerKart");
        model.addAttribute("user", session.getAttribute("user"));
        model.addAttribute("sessionId", session.getId());
        model.addAttribute("authToken", session.getAttribute("auth_token"));
        return "pages/register";
    }

    @PostMapping("/logout")
    public String logout(HttpSession session) {
        session.removeAttribute("auth_token");
        session.removeAttribute("user");
        return "redirect:/login";
    }
}
